[Muhammad Mus'ab] iP

MuhdMusab opened the pull request:
# UncleCheong

> "Hello" - Uncle Cheong ([source](https://www.google.com/search?q=hello))

Uncle Cheong helps you lepak more by helping you with your tasks. It's,

* text-based
* simple to learn
* ~Handsome~ Super handsome

All you need to do is,
1. download it from [here](https://github.com/MuhdMusab/ip/releases/tag/v0.2).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :hot_face:

And it is **FREE!!**

Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
- [ ] World domination (maybe not)

If you're a Java programmer, you can use it to practice Java too. Here's the main method:
```Java
public class Launcher {
    /**
     * Launches the application using the Duke class.
     *
     * @param args
     */
    public static void main(String[] args) {
        try {
            Application.launch(Duke.class, args);
        } catch (Exception e) {
            e.getMessage();
        }
    }
}
```

MuhdMusab committed:
Added Greet,Echo,Exit

MuhdMusab committed:
Added Level 2.Add, list

MuhdMusab committed:
Added Level 2.Add, list

MuhdMusab committed:
Added Level 3. Mark as Done

MuhdMusab committed:
Added Level 4. ToDos, Events, Deadlines

MuhdMusab committed:
Updated message when input is not valid

MuhdMusab committed:
Add automated text ui testing

MuhdMusab committed:
Changed the primitive tasks array to an arraylist

MuhdMusab committed:
Added Level 5. Handle Errors and Command, CommandType and InputParser classes

MuhdMusab committed:
Added Level 6. Delete

MuhdMusab committed:
Updated EXPECTED.TXT

MuhdMusab committed:
Removed extra functions and print statements

MuhdMusab committed:
Updated EXPECTED.TXT and added -w flag to runtest.sh for testing to work properly

MuhdMusab committed:
Updated formatting

MuhdMusab committed:
Level 7: Save

The chatbot allows users to add and modify the Tasks List but does not allow these tasks to be saved and used again.

Let's update the chatbot to allow the saving of tasks into an output file, duke.txt and allow the data from this file to be loaded when Duke starts up.

MuhdMusab committed:
Add Level 8: Dates and Times

The chatbot parses dates/times as Strings and does not convert them to actual dates/times.

Let's allow the chatbot to take in input dates and times when adding deadlines and storing
them as actual dates and times.

MuhdMusab committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Level 7: Save

MuhdMusab committed:
Add Level 8: Dates and Times

MuhdMusab committed:
Revert "Add Level 8: Dates and Times"

This reverts commit f75be430cc14f3545b6bcf8550d8a5614b1cc5d3.

MuhdMusab committed:
Remove empty constructor

MuhdMusab committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Remove empty constructor

MuhdMusab committed:
Merge branch 'master' into branch-Level-7

* master:
  Remove empty constructor
  Revert "Add Level 8: Dates and Times"
  Add Level 8: Dates and Times
  Add Level 8: Dates and Times

MuhdMusab committed:
Merge branch 'branch-Level-7'

* branch-Level-7:

MuhdMusab committed:
Update .gitignore to ignore output

The chatbot produces an output file that is not ignored.

Let's update the .gitignore file to ignore the output

MuhdMusab committed:
Update the reading and writing of stored tasks

The chatbot does not read and write stored tasks correctly.

Let's update the chatbot to read the input by the user and write
the final tasks to an output file.

MuhdMusab committed:
Add Gradle Wrapper

MuhdMusab committed:
Add .class files to .gitignore

MuhdMusab committed:
Add DateTimeParser

Project does not include a class to parse strings of Dates and Times.

Let's include a DateTimeParser to parse strings of Dates and Times to convert the
strings to LocalDateTime objects

MuhdMusab committed:
Add JUnit tests

Project does not include JUnit tests.

Let's include test classes for DateTimeParser, Deadline, Event,
Task and Todo classes

MuhdMusab committed:
Add shadow plugin to Gradle

MuhdMusab committed:
Renamed InputParser to Parser

MuhdMusab committed:
Combine TaskReader and TaskWriter to Storage

MuhdMusab committed:
Add TaskList

MuhdMusab committed:
Add Ui

MuhdMusab committed:
Divide classes into packages

MuhdMusab committed:
Add JavaDoc comments for public methods

MuhdMusab committed:
Update error messages

MuhdMusab committed:
Improve coding standards

MuhdMusab committed:
Add Find command

Chatbot does not allow users to search for their tasks.

Let's include a Find command to allow users to search for
their tasks matching a keyword.

MuhdMusab committed:
Merge tag 'Level-9'

* tag 'Level-9':
  Add Find command

MuhdMusab committed:
Merge tag 'A-JavaDoc'

* tag 'A-JavaDoc':
  Update error messages
  Add JavaDoc comments for public methods

# Conflicts:
#	src/main/java/command/Command.java

MuhdMusab committed:
Merge tag 'A-CodingStandard'

* tag 'A-CodingStandard':
  Improve coding standards

# Conflicts:
#	src/main/java/command/Command.java

MuhdMusab committed:
Update build.gradle to include correct classpath

MuhdMusab committed:
Add checkstyle

MuhdMusab committed:
Fix checkstyle

MuhdMusab committed:
Add unclecheong.jpeg

MuhdMusab committed:
Update gradle version

MuhdMusab committed:
Update gradle version

MuhdMusab committed:
Update ui to include JavaFx functionalities

MuhdMusab committed:
Add JavaFx

MuhdMusab committed:
Merge branch 'Level-10'

* Level-10:
  Add JavaFx

MuhdMusab committed:
Ignore MainWindowBuilder

MuhdMusab committed:
Remove MainWindowBuilder

MuhdMusab committed:
Remove unnecessary icon image

MuhdMusab committed:
Merge tag 'Level-10'

* tag 'Level-10':
  Remove unnecessary icon image
  Remove MainWindowBuilder
  Ignore MainWindowBuilder

MuhdMusab committed:
Update javadocs

MuhdMusab committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Update javadocs

MuhdMusab committed:
Remove unnecessary image

MuhdMusab committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Remove unnecessary image

MuhdMusab committed:
Merge branch 'master' into branch-Level-10

* master:

MuhdMusab committed:
Merge branch 'branch-Level-10'

* branch-Level-10:

aarontangr commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,102 @@
+import ui.MainWindow;
+import ui.Ui;
+
+/**
+ * <h1>Duke class</h1>
+ * Main class of the chat bot that links the
+ * Storage, TaskList and Ui together.
+ */
```
Great usage is <h1>

aarontangr commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,84 @@
+     * @param newDescription to be changed to.
+     */
+    public void setDescription(String newDescription) {
+        this.description = newDescription;
```
No need to use `this.` prefix if not shadowing any variable

aarontangr reviewed (commented):
Looks great. Some cosmetic changes requested.

MuhdMusab committed:
Merge branch 'master' into branch-Level-10

* master:

MuhdMusab committed:
Merge branch 'branch-Level-10'

* branch-Level-10:

MuhdMusab committed:
Remove unnecessary lines

MuhdMusab committed:
Update build.gradle

MuhdMusab committed:
Added fxml files for MainWindow and DialogBox

Project does not support building with fxml files.

Let's include fxml files for MainWindow and DialogBox
to allow building using SceneBuilder

MuhdMusab committed:
Refactor Command methods and constants

Command class does not follow code quality practices.

Let's extract code into methods and refactor code into constants.

MuhdMusab committed:
Improve code quality

garfield-oo7 commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,206 @@
+            throw new DukeException("What talking you");
+        }
+    }
+}
```
This file seems very long, would it be possible to extract all the commands into their own file?

GarethOng replied:
Agreed! A possible solution is to create a Command types (i.e AddCommand, DeleteCommand, Exit Command) that inherit from the Command class to implement their specific behaviour.

garfield-oo7 commented on src/main/java/command/CommandType.java:
```diff
@@ -0,0 +1,19 @@
+ * <h1>CommandType enum</h1>
+ * CommandTypes that can be executed.
+ */
+public enum CommandType {
```
Great use of Enumerations 👍 
!

garfield-oo7 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,63 @@
+                try {
+                    storage.writeToFile(tasks);
+                    ui.sayGoodbye(dialogContainer, userDialog);
+                    PauseTransition delay = new PauseTransition(Duration.seconds(2.5));
```
2.5 seems like a magic number, would it be better to  use a static variable?

garfield-oo7 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,62 @@
+     * @return Command to be executed.
+     */
+    public Command parse(String input) {
+        if (input.equals("")) {
+            return new Command(CommandType.EMPTY, tasks, input, ui);
+        } else if (input.equals("bye")) {
+            return new Command(CommandType.BYE, tasks, input, ui);
+        } else if (input.equals("list")) {
+            return new Command(CommandType.LIST, tasks, input, ui);
+        } else if (input.length() >= 4 && input.substring(0, 4).equals("todo")) {
+            return new Command(CommandType.TODO, tasks, input, ui);
+        } else if (input.length() >= 8 && input.substring(0, 8).equals("deadline")) {
+            return new Command(CommandType.DEADLINE, tasks, input, ui);
+        } else if (input.length() >= 5 && input.substring(0, 5).equals("event")) {
+            return new Command(CommandType.EVENT, tasks, input, ui);
+        } else if (input.length() >= 4 && input.substring(0, 4).equals("mark")) {
+            return new Command(CommandType.MARK, tasks, input, ui);
+        } else if (input.length() >= 6 && input.substring(0, 6).equals("unmark")) {
+            return new Command(CommandType.UNMARK, tasks, input, ui);
+        } else if (input.length() >= 6 && input.substring(0, 6).equals("delete")) {
+            return new Command(CommandType.DELETE, tasks, input, ui);
+        } else if (input.length() >= 4 && input.substring(0, 4).equals("find")) {
+            return new Command(CommandType.FIND, tasks, input, ui);
+        } else {
+            return new Command(CommandType.INVALID, tasks, input, ui);
```
would it better to use switch statements over here to help improve the readability?

GarethOng replied:
Agreed! With switch statements, you can reduce rechecking conditions such as (input.length () >= 6) multiple times.

garfield-oo7 commented on src/main/java/task/Event.java:
```diff
@@ -0,0 +1,38 @@
+     */
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (at: " + DateTimeParser.changeDateTimeFormat(eventDateTime) + ")";
```
I think you can wrap the return statement into the next line, to avoid a coding violation for the number of characters on a line

garfield-oo7 reviewed (commented):
your code is very easy to understand and that you are following the coding standards very closely, I like how you have very well written javadocs for most of the classes and public methods. Some of the classes are very long and you can consider splitting them into smaller classes if possible. Other than that everything else look great 👍

GarethOng commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,206 @@
+        return deadlineDateTime;
+    }
+
+    private boolean hasTaskIndex(int taskIndex) {
+        return taskIndex - 1 < this.tasks.getSize() && taskIndex - 1 >= 0;
+    }
```
Well done! This method can come in handy for different Commands that operates on a specific index in the task list. Implementing it in the parent class is really smart to prevent code duplication!

GarethOng commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    private Task parseLineToTask(String line) throws DukeException {
+        if (line.length() <= 7) {
```
This value of 7 seems out of place. Would it be better to reference it to a static value?

GarethOng commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,38 @@
+     */
+    @Override
+    public String toString() {
+        return "[D]" + super.toString() + " (by: " + DateTimeParser.changeDateTimeFormat(deadlineDateTime) + ")";
```
I believe this violates the coding standard as it exceeds 100 char.

MuhdMusab committed:
Extract the different commands to different classes

MuhdMusab committed:
Update javadocs

MuhdMusab committed:
Add assertions

MuhdMusab committed:
Merge pull request #1 from MuhdMusab/branch-A-CodeQuality

Improve Code Quality

MuhdMusab committed:
Merge branch 'master' into branch-A-Assertions

* master:
  Update javadocs
  Extract the different commands to different classes
  Improve code quality
  Refactor Command methods and constants

# Conflicts:
#	src/main/java/command/Command.java

MuhdMusab committed:
Merge branch 'branch-A-Assertions'

* branch-A-Assertions:
  Add assertions

MuhdMusab committed:
Remove wrong assertions

myangat0343 commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,43 @@
+ * Filters out the commandType and runs the executes the appropriate
+ * command based on the input commandType.
+ */
+public abstract class Command {
```
Good Code!! So well-organized !!!

myangat0343 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,72 @@
+     * @param input String to be parsed.
+     * @return Command to be executed.
+     */
+    public Command parse(String input) {
```
Some suggestions: maybe you can use switch statement for better organization : )

myangat0343 commented on src/main/java/exception/DukeException.java:
```diff
@@ -0,0 +1,30 @@
+package exception;
```
One small suggestions: maybe you could add more sub-exception of DukeException for better organization : )

myangat0343 reviewed (commented):
Your code is good !! Learn a lot from it!

MuhdMusab committed:
Add extension B-ViewSchedule

MuhdMusab committed:
Merge branch 'branch-BCD-Extension'

* branch-BCD-Extension:
  Add extension B-ViewSchedule

MuhdMusab committed:
Add Continuous Integration

MuhdMusab committed:
Update gradle.yml

MuhdMusab committed:
Change ui to include error message

MuhdMusab committed:
Update ui

MuhdMusab committed:
Merge tag 'A-BetterGui'

* tag 'A-BetterGui':
  Update ui

MuhdMusab committed:
Update Todo error handling

MuhdMusab committed:
Merge branch 'A-MoreErrorHandling'

* A-MoreErrorHandling:
  Update Todo error handling

MuhdMusab committed:
Add more JUnit tests

MuhdMusab committed:
Merge branch 'A-MoreTesting'

* A-MoreTesting:
  Add more JUnit tests

MuhdMusab committed:
Add product screenshot

MuhdMusab committed:
Merge branch 'master' into A-MoreErrorHandling

* master:
  Add product screenshot
  Add more JUnit tests

MuhdMusab committed:
Merge tag 'A-MoreErrorHandling'

* tag 'A-MoreErrorHandling':

MuhdMusab committed:
Merge branch 'master' into A-MoreTesting

* master:
  Add product screenshot

MuhdMusab committed:
Merge tag 'A-MoreTesting'

* tag 'A-MoreTesting':

MuhdMusab committed:
Edit .gitignore

MuhdMusab committed:
Merge tag 'A-Personality'

* tag 'A-Personality':
  Edit .gitignore

MuhdMusab committed:
Update build.gradle

MuhdMusab committed:
Update README.md

MuhdMusab committed:
Fix Unmark and Mark command prints and assertions

MuhdMusab committed:
Merge branch 'master' of https://github.com/MuhdMusab/ip

* 'master' of https://github.com/MuhdMusab/ip:
  Update README.md

MuhdMusab committed:
Refactor Junit tests

MuhdMusab committed:
Remove whitespace

MuhdMusab committed:
Remove javafx tag

MuhdMusab committed:
Update README.md

MuhdMusab committed:
Update documentation for Ui

MuhdMusab committed:
Update README.md

MuhdMusab committed:
Update README.md

MuhdMusab committed:
Change ViewScheduleCommand method name

MuhdMusab committed:
Merge remote-tracking branch 'origin/master'

MuhdMusab committed:
Update xmlns of DialogBox

MuhdMusab committed:
Update README.md

MuhdMusab committed:
Update README.md
